runtime/lua/vim/filetype/detect.lua

```diff
@@ -1204,4 +1286,269 @@ end
 -- luacheck: pop
 -- luacheck: pop
 
+local patterns_hashbang = {
+  ['^zsh\\>'] = { 'zsh', { vim_regex = true } },
+  ['^\\(tclsh\\|wish\\|expectk\\|itclsh\\|itkwish\\)\\>'] = { 'tcl', { vim_regex = true } },
+  ['^expect\\>'] = { 'expect', { vim_regex = true } },
+  ['^gnuplot\\>'] = { 'gnuplot', { vim_regex = true } },
+  ['make\\>'] = { 'make', { vim_regex = true } },
+  ['^pike\\%(\\>\\|[0-9]\\)'] = { 'pike', { vim_regex = true } },
+  lua = 'lua',
+  perl = 'lua',
+  php = 'php',
+  python = 'python',
+  ['^groovy\\>'] = { 'groovy', { vim_regex = true } },
+  raku = 'raku',
+  ruby = 'ruby',
+  ['node\\(js\\)\\=\\>\\|js\\>'] = { 'javascript', { vim_regex = true } },
+  ['rhino\\>'] = { 'javascript', { vim_regex = true } },
+  -- BC calculator
+  ['^bc\\>'] = { 'bc', { vim_regex = true } },
+  ['sed\\>'] = { 'sed', { vim_regex = true } },
+  ocaml = 'ocaml',
+  -- Awk scripts; also finds "gawk"
+  ['awk\\>'] = { 'awk', { vim_regex = true } },
+  wml = 'wml',
+  scheme = 'scheme',
+  cfengine = 'cfengine',
+  escript = 'erlang',
+  haskell = 'haskell',
+  clojure = 'clojure',
+  ['scala\\>'] = { 'scala', { vim_regex = true } },
+  -- Free Pascal
+  ['instantfpc\\>'] = { 'pascal', { vim_regex = true } },
+  ['fennel\\>'] = { 'fennel', { vim_regex = true } },
+  -- MikroTik RouterOS script
+  ['rsc\\>'] = { 'routeros', { vim_regex = true } },
+  ['fish\\>'] = { 'fish', { vim_regex = true } },
+  ['gforth\\>'] = { 'forth', { vim_regex = true } },
+  ['icon\\>'] = { 'icon', { vim_regex = true } },
+}
+
+---@private
+-- File starts with "#!".
+local function match_from_hashbang(contents, path)
+  local first_line = contents[1]
+  -- Check for a line like "#!/usr/bin/env {options} bash".  Turn it into
+  -- "#!/usr/bin/bash" to make matching easier.
+  -- Recognize only a few {options} that are commonly used.
+  if M.matchregex(first_line, [[^#!\s*\S*\<env\s]]) then
```

Comment:
```suggestion
  if matchregex(first_line, [[^#!\s*\S*\<env\s]]) then
```
leftover from moving this back and forth :D